### ISSUE TITLE
[QUESTION] Reset trait on edit the content of a custom component

### ISSUE THREAD

**Issue by inaLar** [artf/grapesjs#2193]
Hi,

I have a custom component which content can be changed by switching trait select options, but also can be edited similar to the text component type when it is already rendered inside the canvas.
Is there an easy way to reset the traits related attribute on edit of a component?

`model: {
        defaults:  {
         type: 'tag',
          tagName: 'span',
          droppable: 0,
          editable: 1,
          textable: 1,
          traits: [{
            type: 'select',
            label: 'Custom type',
            name: 'customType',
            options: customTypeOptions,
          }],
        },
toHTML() {
          return this.get('content');
        },`

`view: {
 events: {
            focus: 'resetTrait',
	},
reseTrait() {
}
}`

**Comment by Zachary1748** [comment]
You could attach an event listener to the content prop of the component, then in the callback fetch the trait or change the value prop of the trait, see example below [taken from docs](https://grapesjs.com/docs/modules/Traits.html#updating-traits-at-run-time)

```javascript
// Let's update `options` of our `type` trait, defined in Input component
const component = editor.getSelected();
component.getTrait('type').set('options', [
  { id: 'opt1', name: 'New option 1'},
  { id: 'opt2', name: 'New option 2'},
]);
// or with multiple values
component.getTrait('type').set({
  label: 'My type',
  options: [...],
});
```

**Comment by inaLar** [comment]
Thank you Zachary1748, but I can't use that in my case. Maybe I didn't explained it right.
I have a custom made component. 
The components content can be changed in 3 ways:
1. When dragged and droped into the canvas (you have a popup with options)
2. From Settings (also via choosing between options)
3. By editing its text inside the canvas

Anyway I think I fount a solution - I created a custom data type and binded the events to its change also, not only to the content changes. But I am still testing it.

**Comment by Zachary1748** [comment]
@inaLar Would you like to elaborate on why you can't use the standard method? After dragging a block into the canvas it retrieves the appropriate component type based on the contents of the block. Once you have a model instance of the component, it should be fairly straight forward using the method I referenced in any of your scenarios; to either update the component models' attribute value which is binded to the trait, or update the trait options based on your requirements, both of which may be triggered from events.

**Comment by inaLar** [comment]
I have done this + other things - like, the component can be dragged in and out of a text block, and also can be edited by typing, when the  value of the attribute, related to the trait is changed again. So I believe I've handled every flow for now. I will close the question.

**Comment by Zachary1748** [comment]
Glad you got it handled! Yes I agree, @artf could use a bit of work on the docs to provide a byte of better explanations and use cases.

**Comment by artf** [comment]
> Just one more thing - my trait is "Select", so on marking the custom element in the canvas, the settings correspond to the current value of the components content. (this is working perfectly)

@inaLar I didn't get the problem here

> And maybe the documentation is more difficult for me to use.

> Yes I agree, @artf could use a bit of work on the docs to provide a byte of better explanations and use cases.

In the last release, I rewrote the Trait documentation from scratch (and added new API for customization) by trying to cover all the possible needs requested so far by GrapesJS users. Obviously, I can't cover all cases for all the people but if you have any suggestion are more than welcome

**Comment by inaLar** [comment]
@artf, regarding the "Select" I couldn't manage to connect the change in the content and make it work on change of the trait in one flow and also to change it manually by editing it's text/content. The manual change wasn't syncing the trait, but I managed to make it work. A custom text component worked just fine.

As for the documentation from the last release, I am reading it right now and it looks very good, but I haven't seen it when I posted the question.

It may be only me, but I learn by example, so more examples and use cases will be always a plus.

I understand that the product is open and I appreciate your effort. I hope that I will gain more knowledge and will be able to contribute with something in near future.

Cheers.
